pmd-core/src/main/java/net/sourceforge/pmd/lang/ast/TokenMgrError.java

```diff
@@ -0,0 +1,123 @@
+
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import net.sourceforge.pmd.annotation.InternalApi;
+import net.sourceforge.pmd.util.StringUtil;
+
+/**
+ * An error thrown during lexical analysis of a file.
+ */
+public final class TokenMgrError extends RuntimeException {
+
+
+    /**
+     * @deprecated Will be removed when all modules are ported
+     */
+    @Deprecated
+    public static final int LEXICAL_ERROR = 0;
+
+    /**
+     * @deprecated Will be removed when all modules are ported,
+     *     see {@link #TokenMgrError(String, int)}
+     */
+    @Deprecated
+    public static final int INVALID_LEXICAL_STATE = 1;
+
+    private final int line;
+    private final int column;
+    private final String filename;
+
+    public TokenMgrError(int line, int column, @Nullable String filename, String message, @Nullable Throwable cause) {
+        super(message, cause);
+        this.line = line;
+        this.column = column;
+        this.filename = filename;
+    }
+
+    public TokenMgrError(int line, int column, String message, @Nullable Throwable cause) {
+        this(line, column, null, message, cause);
+    }
+
+    /**
+     * @deprecated This is used by javacc but those usages are being replaced with an IllegalArgumentException
+     */
+    @Deprecated
+    @SuppressWarnings("PMD.UnusedFormalParameter")
+    public TokenMgrError(String message, int errorCode) {
+        this(-1, -1, null, message, null);
+    }
+
+    /**
+     * Constructor called by JavaCC.
+     */
+    @InternalApi
+    @SuppressWarnings("PMD.UnusedFormalParameter")
+    public TokenMgrError(boolean eofSeen, int lexState, int errorLine, int errorColumn, String errorAfter, char curChar) {
+        super(makeMessage(eofSeen, errorLine, errorColumn, errorAfter, curChar));
+        line = errorLine;
+        column = errorColumn;
+        filename = AbstractTokenManager.getFileName();
+    }
+
+    /**
+     * Constructor called by JavaCC.
+     *
+     * @deprecated The error code is useless, ported modules use the other constructor
+     */
+    @Deprecated
+    @SuppressWarnings("PMD.UnusedFormalParameter")
+    public TokenMgrError(boolean eofSeen, int lexState, int errorLine, int errorColumn, String errorAfter, char curChar, int errorCode) {
+        super(makeMessage(eofSeen, errorLine, errorColumn, errorAfter, curChar));
+        line = errorLine;
+        column = errorColumn;
+        filename = AbstractTokenManager.getFileName();
+    }
+
+    public int getLine() {
+        return line;
+    }
+
+    public int getColumn() {
+        return column;
+    }
+
+    public @Nullable String getFilename() {
+        return filename;
+    }
+
+    /**
+     * Returns a detailed message for the Error when it is thrown by the
+     * token manager to indicate a lexical error.
+     * Parameters :
+     * eofseen     : indicates if EOF caused the lexical error
+     * curLexState : lexical state in which this error occurred
```

Comment:
Not many languages really use different lexical states. Though maybe, we can use the parameter later anyway. We'd need a callback to get `lexStateNames`. Maybe, we should define some "master" interface, which provides access to all this static info (lexical state names, node names, token names, parser constructor, token document constructor, etc). Or maybe just an AbstractJavaccParser would suffice.

Reply:
This is done now. I modified the calls to the constructor so that it passes the lexical state name instead of the number. This happens only for modules that use the new ant build.